### DISCUSSION TITLE
best-practices/accelerating-dotfiles-debugging

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#168] [Announcements]
# best-practices/accelerating-dotfiles-debugging

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/best-practices/accelerating-dotfiles-debugging

**Comment by CUexter** [comment]
why not just use mkoutofstoresymlink ?

**Reply by ryan4yin** [reply]
I just found out that such a function even exists! I had to search high and low before finally locating its definition and usage in the [Home Manager source code](https://github.com/search?q=repo%3Anix-community%2Fhome-manager%20outOfStoreSymlink&type=code). I'm truly amazed...

This function is indeed much simpler than my approach. Thank you so much for the heads-up!

**Comment by nanoni17728** [comment]
Hi Ryan! Thank you for all your hard work on the book. It made me able to push beyond the basics of NixOS :)

I think this page lacks some context that the less attentives (eg. me) might miss. I had to clear two points before making `mkOutOfStoreSymlink` work.

The first one is easy as it is already briefly mentioned and is about the basics of flakes. The paths to the files *have* to be absolute, and passed as strings at that, since we are working on a flake-based configuration and flakes can't reach outside of their directory. Right now this isn't detailed and can make the configuration look like an arbitrary choice at a glance. It feels like one of those cases where we could assume the reader has enough knowledge to understand the implications by themselve (and read the function) but for the amount of text it requires making sure everyone is on the same page would be a good safety net.

The second one is not as trivial and would require some more substancial writing. I am still mentioning it in the hope it can help others! `config` in this scenario as I understand it is the `config` that is passed on by home-manager. Which means, you have to be in the context of the home-manager module. If, like me, you lumped together the NixOS user definition and its home-manager conterpart in the same file, nix will throw hands at you since the `config` you get passed on is the NixOS one (my guess). You will have to separate the two, then choose a way to correctly import the home-manager part. [The way the book shows it](https://nixos-and-flakes.thiscute.world/nixos-with-flakes/start-using-home-manager) is to import the file while loading the module in `flake.nix`.

[This reddit thread](https://www.reddit.com/r/NixOS/comments/vh2kf7/home_manager_mkoutofstoresymlink_issues/) showed me that `home-manager.users.username` also have an `imports` attribute. If you're really adamant on keeping user definition tightly coupled in a corner (as I am) you can make a `user/` directory with a `default.nix` that defines your NixOS user and import your `home.nix`. [See there](https://gist.github.com/nanoni17728/2ba2588fe65e2eee6140fb7654de57e2).

As a nice touch aliasing `config.lib.file.mkOutOfStoreSymlink` to `mkOutOfStoreSymlink` in the `let` binding could make things pretty ✨

Hope this isn't too much of a wall of text. Have a nice day!
